In [15]:
!conda install -c conda-forge beautifulsoup4 --yes

!conda install -c conda-forge geopy --yes

!conda install -c conda-forge folium=0.5.0 --yes

print('Libraries installed.')

Solving environment: done

## Package Plan ##

  environment location: /anaconda3

  added / updated specs:
    - beautifulsoup4


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2018.4.16          |           py36_0         142 KB  conda-forge
    conda-4.6.14               |           py36_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.2 MB

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi                                         pkgs/main --> conda-forge
  conda                                           pkgs/main --> conda-forge



certifi-2018.4.16    | 142 KB    | ##################################### | 100% 
conda-4.6.14         | 2.1 MB    | ##################################### | 100% 
Preparing transaction: done
Verifying transactio

In [17]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import requests
from pandas.io.json import json_normalize
import json

import requests

from bs4 import BeautifulSoup

from geopy.geocoders import Nominatim

import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

print('Libraries imported.')


Libraries imported.


**Part 1. Data Scraping: 
Build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe.
 - The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
 - Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
 - More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.
- If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
- Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
- In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.**

**We will now scrape the Wikipedia page. 
The one table on this page has 3 columns: postal code, borough and neighborhood. Postal codes start with the letter M.**

In [18]:
# open Wiki page with Beautiful Soup
wikipedia_data = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(wikipedia_data, 'html.parser')

In [19]:
postalcodeList = []
boroughList = []
neighborhoodList = []

for row in soup.find('table').find_all('tr'):
    cells = row.find_all('td')
    if(len(cells) > 0):
        postalcodeList.append(cells[0].text)
        boroughList.append(cells[1].text)
        neighborhoodList.append(cells[2].text.rstrip('\n')) 
        
# remove the new line from neighborhood cell

**The 3 lists from the wikipedia page will now be put in a dataframe.**

In [20]:
toronto_neighorhood = [('PostalCode', postalcodeList),
                      ('Borough', boroughList),
                      ('Neighborhood', neighborhoodList)]
toronto_df = pd.DataFrame.from_dict(dict(toronto_neighorhood))
toronto_df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

**Remove rows without a borough assigned.**

In [21]:
toronto_df_dropna = toronto_df[toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
toronto_df_dropna.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

 **concat neighborhoods which have same borough into same row, seperated by a comma.**

In [22]:
toronto_df_grouped = toronto_df_dropna.groupby(['PostalCode','Borough'], as_index=False).agg(lambda x: ','.join(x))
toronto_df_grouped.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

**Replace the 'Not assigned' in Neighborhood with the corresponding Borough**

In [23]:
na_neigh_rows = toronto_df_grouped.Neighborhood == 'Not assigned'
toronto_df_grouped.loc[na_neigh_rows, 'Neighborhood'] = toronto_df_grouped.loc[na_neigh_rows, 'Borough']
toronto_df_grouped[na_neigh_rows]

PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Queen's Park

In [24]:
toronto_df_final = toronto_df_grouped
toronto_df_final.shape

(103, 3)

**final dataframe has 103 rows**

**Part 2: Adding geo coordinates to the dataframe**

In [25]:
latlongurl="http://cocl.us/Geospatial_data"
coordinates=pd.read_csv(latlongurl)

In [11]:
print(coordinates.shape)
coordinates.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

**Latitude and Longitude file also has 103 rows as 103 postal codes in the Toronto dataframe**

**Next step - Merge: set indexes of the two dataframes to the Postal Code columns**

In [26]:
toronto_df_temporary = toronto_df_final.set_index('PostalCode')
coordinates_temporary = coordinates.set_index('Postal Code')
toronto_df_coordinates = pd.concat([toronto_df_temporary, coordinates_temporary], axis=1, join='inner')

**Reset index**

In [28]:
toronto_df_coordinates.index.name = 'PostalCode'
toronto_df_coordinates.reset_index(inplace=True)

**Check the merged dataframe**

In [29]:
print(toronto_df_coordinates.shape)
toronto_df_coordinates.head()

(103, 5)


PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

**Part 3: Explore and cluster the neighborhoods in Toronto**

In [30]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


**create a Map of Toronto with Boroughs markers**

In [34]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coordinates['Latitude'], toronto_df_coordinates['Longitude'], toronto_df_coordinates['PostalCode'], toronto_df_coordinates['Borough'], toronto_df_coordinates['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=7,
        popup=popup,
        color='red',
        fill=True,
        fill_color='#cc315c',
        fill_opacity=0.6,
        parse_html=False).add_to(map_toronto)
    
map_toronto

**Reduce the number of Boroughs:  To reduce the numbers of calls to FourSquare API, let's explore boroughs that have Toronto in their names.**

In [32]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coordinates[toronto_df_coordinates['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(38, 5)


PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [33]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=7,
        popup=popup,
        color='red',
        fill=True,
        fill_color='#cc315c',
        fill_opacity=0.6,
        parse_html=False).add_to(map_toronto)
    
map_toronto


**Use FourSquare API to explore the Boroughs**

**Get recommend places in Toronto Central**



In [36]:
CLIENT_ID = '3ZAHRAPDXGU2GFBUXUMVMOKZUM24ATFAX2YXRDX3SIM1FYYM' # Foursquare ID
CLIENT_SECRET = 'WNAKD3SRFVCUVFYNSDCYDTRGYBPIA3YMY31QZYEV13Q53DOT' # Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [37]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [38]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(1701, 9)


PostalCode       Borough                 Neighborhood  BoroughLatitude  \
0        M4E  East Toronto                  The Beaches        43.676357   
1        M4E  East Toronto                  The Beaches        43.676357   
2        M4E  East Toronto                  The Beaches        43.676357   
3        M4E  East Toronto                  The Beaches        43.676357   
4        M4K  East Toronto  The Danforth West,Riverdale        43.679557   

   BoroughLongitude                           VenueName  VenueLatitude  \
0        -79.293031  The Big Carrot Natural Food Market      43.678879   
1        -79.293031                 Grover Pub and Grub      43.679181   
2        -79.293031               St-Denis Studios Inc.      43.675031   
3        -79.293031                       Upper Beaches      43.680563   
4        -79.352188                            Pantheon      43.677621   

   VenueLongitude      VenueCategory  
0      -79.297734  Health Food Store  
1      -79.297215                Pub  
2      -79.288022        Music Venue  
3      -79.292869       Neighborhood  
4      -79.351434   Greek Restaurant

**display count of venues in each postal code**

In [39]:
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()


PostalCode  Borough           Neighborhood                                                                                        
M4E         East Toronto      The Beaches                                                                                               4
M4K         East Toronto      The Danforth West,Riverdale                                                                              44
M4L         East Toronto      The Beaches West,India Bazaar                                                                            22
M4M         East Toronto      Studio District                                                                                          37
M4N         Central Toronto   Lawrence Park                                                                                             4
M4P         Central Toronto   Davisville North                                                                                          7
M4R         Central Toronto   North Toron

**Check number of venue categories returned**

In [41]:
len(venues_df['VenueCategory'].unique())

236

**Analyze venues**

In [42]:
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood columns back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1701, 239)


PostalCode       Borough                Neighborhoods  Adult Boutique  \
0        M4E  East Toronto                  The Beaches               0   
1        M4E  East Toronto                  The Beaches               0   
2        M4E  East Toronto                  The Beaches               0   
3        M4E  East Toronto                  The Beaches               0   
4        M4K  East Toronto  The Danforth West,Riverdale               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Argentinian Restaurant  Art Gallery  Art Museum  \
0             0         0                       0            0           0   
1             0         0                       0            0           0   
2             0         0                       0            0           0   
3             0         0                       0            0           0   
4             0         0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0      0         0           0                   0   
1                   0      0         0           0                   0   
2                   0      0         0           0                   0   
3                   0      0         0           0                   0   
4                   0      0         0           0                   0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0        

**Get the frequency of occurance of each category in an area**

In [44]:
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()

(38, 239)


PostalCode          Borough                  Neighborhoods  Adult Boutique  \
0        M4E     East Toronto                    The Beaches             0.0   
1        M4K     East Toronto    The Danforth West,Riverdale             0.0   
2        M4L     East Toronto  The Beaches West,India Bazaar             0.0   
3        M4M     East Toronto                Studio District             0.0   
4        M4N  Central Toronto                  Lawrence Park             0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.022727   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.054054   
4             0.0              0.0               0.0             0.000000   

   Antique Shop  Aquarium  Argentinian Restaurant  Art Gallery  Art Museum  \
0           0.0       0.0                     0.0          0.0         0.0   
1           0.0       0.0                     0.0          0.0         0.0   
2           0.0       0.0                     0.0          0.0         0.0   
3           0.0       0.0                     0.0          0.0         0.0   
4           0.0       0.0                     0.0          0.0         0.0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                  0.0               0.0                 0.0            0.0   
1                  0.0               0.0                 0.0            0.0   
2                  0.0               0.0                 0.0            0.0   
3                  0.0               0.0                 0.0            0.0   
4                  0.0               0.0                 0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop    Bakery      Bank       Bar  \
0        0.0         0.0         0.0  0.000000  0.000000  0.000000   
1        0.0         0.0         0.0  0.022727  0.000000  0.000000   
2        0.0         0.0         0.0  0.000000  0.000000  0.000000   
3        0.0         0.0         0.0  0.054054  0.027027  0.027027   
4        0.0         0.0         0.0  0.000000  0.000000  0.000000   

   Baseball Stadium  Basketball Stadium  Beach  Beer Bar  Beer Store  \
0               0.0                 0.0    0.0       0.0         0.0   
1               0.0                 0.0    0.0       0.0         0.0   
2               0.0                 0.0    0.0       0.0         0.0   
3               0.0                 0.0    0.0       0.0         0.0   
4               0.0                 0.0    0.0       0.0         0.0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                 0.0     0.0            0.0   0.000000       0.0   
1                 0.0     0.0            0.0   0.045455       0.0   
2                 0.0     0.0            0.0   0.000000       0.0   
3                 0.0     0.0            0.0   0.027027       0.0   
4                 0.0     0.0            0.0   0.000000       0.0   

   Brazilian Restaurant  Breakfast Spot   Brewery  Bubble Tea Shop  Building  \
0                   0.0             0.0  0.000000         0.000000       0.0   
1                   0.0             0.0  0.022727         0.022727       0.0   
2                   0.0             0.0  0.045455         0.000000       0.0   
3                   0.0             0.0  0.027027         0.000000       0.0   
4                   0.0             0.0  0

**10 most frequented venue types**

In [45]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
35        M6R      West Toronto   
3         M4M      East Toronto   
34        M6P      West Toronto   
36        M6S      West Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
6         M4R   Central Toronto   
19        M5J  Downtown Toronto   
21        M5L  Downtown Toronto   
20        M5K  Downtown Toronto   
29        M5X  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
14        M5B  Downtown Toronto   
16        M5E  Downtown Toronto   
15        M5C  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
11        M4X  Downtown Toronto   
9         M4V   Central Toronto   
28        M5W  Downtown Toronto   
24        M5R   Central Toronto   
22        M5N   Central Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
5         M4P   Central Toronto   
37        M7Y      East Toronto   
0         M4E      East Toronto   
4         M4N   Central Toronto   
10        M4W  Downtown Toronto   
2         M4L      East Toronto   
31        M6H      West Toronto   
8         M4T   Central Toronto   
7         M4S   Central Toronto   
23        M5P   Central Toronto   

                                        Neighborhoods 1st Most Common Venue  \
27  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   
32                            Little Portugal,Trinity                   Bar   
33         Brockton,Exhibition Place,Parkdale Village        Breakfast Spot   
35                              Parkdale,Roncesvalles        Breakfast Spot   
3                                     Studio District                  Café   
34                       High Park,The Junction South                  Café   
36                                  Runnymede,Swansea                  Café   
25                      Harbord,University of Toronto                  Café   
26            Chinatown,Grange Park,Kensington Market                  Café   
6                                  North Toronto West        Clothing Store   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
21                      Commerce Court,Victoria Hotel           Coffee Shop   
20            Design Exchange,Toronto Dominion Centre           Coffee Shop   
29              First Canadian Place,Underground city           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
18                             Adelaide,King,Richmond           Coffee Shop   
14                            Ryerson,Garden District           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
15                                     St. James Town           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
13                           Harbourfront,Regent Park           Coffee Shop   
11                         Cabbagetown,St. James Town           Coffee Shop   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...           Coffee Shop   
28                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
24                  The Annex,North Midtown,Yorkville           Coffee Shop   
22                                           Roselawn                Garden   
1                         The Danforth West,Riverdale      Greek Restaurant   
30                                           Christie         Grocery Store   
5                                    Davisville North                   Gym   
37  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
0                                         The Beaches           Music Venue   
4                                       Lawrence Park                  Park   
10                            

**Clustering areas: 
Use KMeans into 3 clusters**

In [46]:
kclusters = 3

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
10        M4W  Downtown Toronto   
8         M4T   Central Toronto   
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
35        M6R      West Toronto   
3         M4M      East Toronto   
34        M6P      West Toronto   
36        M6S      West Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
6         M4R   Central Toronto   
19        M5J  Downtown Toronto   
21        M5L  Downtown Toronto   
20        M5K  Downtown Toronto   
29        M5X  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
14        M5B  Downtown Toronto   
16        M5E  Downtown Toronto   
15        M5C  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
11        M4X  Downtown Toronto   
9         M4V   Central Toronto   
28        M5W  Downtown Toronto   
24        M5R   Central Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
5         M4P   Central Toronto   
37        M7Y      East Toronto   
0         M4E      East Toronto   
4         M4N   Central Toronto   
2         M4L      East Toronto   
31        M6H      West Toronto   
7         M4S   Central Toronto   
23        M5P   Central Toronto   
22        M5N   Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
10                                           Rosedale  43.679563 -79.377529   
8                          Moore Park,Summerhill East  43.689574 -79.383160   
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420   
32                            Little Portugal,Trinity  43.647927 -79.419750   
33         Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191   
35                              Parkdale,Roncesvalles  43.648960 -79.456325   
3                                     Studio District  43.659526 -79.340923   
34                       High Park,The Junction South  43.661608 -79.464763   
36                                  Runnymede,Swansea  43.651571 -79.484450   
25                      Harbord,University of Toronto  43.662696 -79.400049   
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049   
6                                  North Toronto West  43.715383 -79.405678   
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752   
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817   
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576   
29              First Canadian Place,Underground city  43.648429 -79.382280   
17                                 Central Bay Street  43.657952 -79.387383   
18                             Adelaide,King,Richmond  43.650571 -79.384568   
14                            Ryerson,Garden District  43.657162 -79.378937   
16                                        Berczy Park  43.644771 -79.373306   
15                                     St. James Town  43.651494 -79.375418   
12                               Church and Wellesley  43.665860 -79.383160   
13                           Harbourfront,Regent Park  43.654260 -79.360636   
11                         Cabbagetown,St. James Town  43.667967 -79.367675   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
24                  The Annex,North Midtown,Yorkville  43.672710 -79.405678   
1                         The Danforth West,Riverdale  43.679557 -79.352188   
30                                           Christie  43.669542 -79.422564   
5                                    Davisville North  43.712751 -79.390197   
37  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558   
0                                         The Beaches  43.676357 -79.293031   
4                             

**map the clusters**

In [48]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Observations about the clusters:

Cluster 0 (red): Residential, with day to day amenities for residents
Cluster 1 (blue): variety of venues, particularly cafes
Cluster 2 (green): Market, with small shops and restaurants